Update Bleak Fork to Upstream version

pmessan opened the pull request:
In order to reduce our tech debt and improve maintainability of our code, I propose a shift to the upstream version of Bleak. Fortunately there is not much to be patched, and functionality has been preserved. This has been tested with [this progateway PR](https://github.com/workaroundgmbh/progateway/pull/982) and [this problew PR](https://github.com/workaroundgmbh/progateway/pull/982).

es617 committed:
Adding mtu_size client property (#525)

dlech committed:
remove unnecessary "is True"

dlech committed:
CoreBluetooth: restore waiting for completion in notification start/stop

This was commented out in f3bead73bc044f1eace4ed77672ad46cc66318ea
without explanation and seems unrelated to that change.

dlech committed:
Fix type hints in CoreBluetooth PeripheralDelegate

Unfortunately, NSArray type does not support subscripting for the element type.

dlech committed:
Fixed write without response on BlueZ < 5.51.

This is a regression from when we switched to dbus-next. It doesn't
return the fds as part of the body but rather in a separate attribute.

dlech committed:
Add hack for getting MTU on BlueZ backend.

BlueZ doesn't have a good way to get the negotiated MTU, but we can get
it by abusing existing D-Bus methods.

Since the workaround is async and could cause unwanted side effects for
some devices, it opt-in as seen in the example code.

dlech committed:
Merge pull request #542 from dlech/bug/dbus-next-unix-fd

Fixed write without response on BlueZ < 5.51.

dlech committed:
Merge branch 'develop' into mac-fixes

dlech committed:
Merge branch 'develop' into change_typing_of_data_parameter

dlech committed:
include memoryview in bytes-like type hints

dlech committed:
Merge pull request #520 from sander76/change_typing_of_data_parameter

change typing on data parameter in `write_gatt_char` and `write_gatt_descriptor`

devanlai committed:
winrt: report rssi

Report RSSI in the standard BLEDevice `rssi` field added in #390

dlech committed:
Merge pull request #544 from devanlai/winrt-rssi

Populate rssi field for WinRT backend

dlech committed:
Merge pull request #543 from dlech/feature/bluez-mtu

Add hack for getting MTU on BlueZ backend.

dlech committed:
Merge pull request #541 from dlech/mac-fixes

Mac fixes

dlech committed:
Add CI environment variable check to import.

Using Bleak on CI will fail if `bluetoothctl` is not installed. This
adds a check for the CI environment variable that is set by default in
many CI environments (e.g. GitHub Actions, Travis CI, etc.). Since we
aren't sure that the value will be the same on all CI platforms, we just
check for the presence of the variable. The check extends the same
workaround that is already in place for building on readthedocs.org.

Commit:
Fix propagation of CoreBluetooth errors

Errors from CBPeripeheralDelegate calls were being raised in the asyncio
event loop, but not in the correct task. In order to reach the user's
code, the error has to be passed through the event to the user call that
is awaiting the event.

Without these changes, e.g. ATT errors sent by the peripheral caused
error messages to be printed, but since the event that the user's call
was waiting on was never set, the user's code was blocked.

The implementation was adapted from https://github.com/hbldh/bleak/pull/209

dlech committed:
Merge pull request #554 from rtoijala/macos-errors

Fix propagation of CoreBluetooth errors

dlech committed:
Merge branch 'develop' into feature/ci

dlech committed:
Merge pull request #552 from dlech/feature/ci

Add CI environment variable check to import.

dlech committed:
Fixed notification value should be bytearray on CoreBluetooth.

This fixes the value passed to the notification callback not being cast
to bytearray like other backends.

Fixes #560

dlech committed:
Improved type hints in CoreBluetooth backend.

This adds and fixes many type hints in the CoreBluetooth backend and
includes .pyi typings for the PyObjC types.

The `typings/` folder is the standard place for Pyright to find the typings.

dlech committed:
Removed use of CMDConnectionState in CoreBluetooth backend.

This removes the peripheral state management from the central manager.
The CBPeripheral objects keep track of their own state which can be
fetched via the state() property, so we don't need to keep track of the
state separately.

The use of the state change event is replaced with futures. Also, this
change now properly waits for a callback after cancelling the connection
in the event of a asyncio.TimeoutError which happens to fix #538.

dlech committed:
Fix connecting to multiple devices on Mac

This fixes connecting to multiple devices when using a single
BleakScanner in the CoreBluetooth backend. Previously, the central
manager held a reference to a single peripheral, so if a second
peripheral was connected using the same central manager, it would
overwrite the first.

This removes the coupling so that the peripheral is passed as an
argument to the central manager methods instead.

dlech committed:
simplify CoreBluetooth mtu_size()

This removes the extra call to the PeripheralDelegate when getting the
MTU size on macOS. Since we don't need to wait for a callback, there
isn't a need to have the extra function in the delegate.

dlech committed:
Use delegate callbacks for get_rssi() on CoreBluetooth backend.

This changes the BleakClientCoreBluetooth.get_rssi() method to use
the available delegate callbacks in the peripheral delegate. This
will ensure that we don't return stale RSSI values and will pass
errors as exceptions instead of just returning None.

dlech committed:
Fixed deadlock in CoreBluetooth backend when device disconnects

This fixes #535. In the CoreBluetooth backend, when a device disconnects,
none of the peripheral delegate callbacks will be called. So if any
futures are pending, they will be waiting forever. This adds a disconnect
hook that will raise an exception in all pending futures to avoid this
deadlock.

dlech committed:
Fix handling of futures for same UUID on CoreBluetooth

This fixes potential deadlocks when using a service, characteristic or
descriptor with the same UUID on the CoreBluetooth backend. Previously,
the futures were stored by UUID for later lookup in callbacks. However,
since since the same UUID can be used multiple times, we could end up
with futures being replaced in a dictionary before they were completed,
which would result in a deadlock for the waiter.

dlech committed:
drop string2uuid helper function

This function was only used once and it is just as simple to use the
wrapped function directly.

dlech committed:
drop scanForPeripherals_ method

This was not used anywhere.

dlech committed:
Use @objc.python_method where possible in PeripheralDelegate class.

These methods are only used in Python, so don't need to be Obj-C selectors.

jackjansen committed:
Added option command line argument (mac-address or uuid of device to scan for)

jackjansen committed:
Added find_device_by_filter() which searches devices by filter function. Added example that uses this functionality to find devices by name.

jackjansen committed:
Added typing info

jackjansen committed:
Ran black on the code

jackjansen committed:
Added optional command line arguments to specify device and other parameters on the command line

jackjansen committed:
Added changelog entries

jackjansen committed:
Use a queue to communicate found BLEDevice to find_device_by_filter

hbldh committed:
Adding a stop event in WinRT scanner

dlech committed:
remove unused import

dlech committed:
drop stop_scanning_event in find_device_by_filter

Since we can await the asyncio.Queue, we no longer need an asyncio.Event().

Also remove unused timeout kwarg while we are touching this code.

dlech committed:
Merge pull request #565 from jackjansen/discovery-filter

Discovery filter

dlech committed:
Merge pull request #561 from dlech/bug/cb-notify-bytearray

Fixed notification value should be bytearray on CoreBluetooth.

dlech committed:
Merge branch 'develop' into mac-cleanup-and-fixes

dlech committed:
Merge pull request #564 from dlech/mac-cleanup-and-fixes

Mac cleanup and fixes

dlech committed:
CoreBluetooth scan start/stop improvements

This makes use of ObjC-style key-value observer to monitor the isScanning
property on CBCentralManager. This allows us to avoid the polling wait.

Previously, the wait was only implemented when stopping scanning, so it
is added to the start as well.

Also, the higher-level function is fixed to allow calling stop()
multiple times without raising an error. This is needed since the
stop() method can be called by __aexit__() even if the scanner is
already stopped.

hbldh committed:
Implemented sstop event for pythonnet backend as well.
Wrote details in CHANGELOG.rst

hbldh committed:
Merge branch 'develop' into feature/stop-event-fix

hbldh committed:
Merge pull request #571 from hbldh/feature/stop-event-fix

dlech committed:
Merge pull request #572 from dlech/mac-start-stop-scan

CoreBluetooth scan start/stop improvements

dlech committed:
Release v0.12.0

Update changelog and version for release.

dlech committed:
github: drop bluez install on CI

This should no longer be needed since 84af9f3.

dlech committed:
Merge pull request #575 from hbldh/release/0.12.0

Release v0.12.0

dlech committed:
Merge pull request #578 from hbldh/master

Merge release back into develop

dlech committed:
bump version for develop

dlech committed:
CHANGELOG: fix typos

dlech committed:
remove unused listen_properties_changed function

dlech committed:
CoreBluetooth: fix unawaited future for write without response

In CoreBluetooth, there is no feedback for write without response, so
we just have to send it and hope for the best. In this case we should
not create a future since it gets registered with the event loop but it
is never awaited.

Fixes #586

dlech committed:
Merge pull request #588 from hbldh/mac-write-wo-response-future

CoreBluetooth: fix unawaited future for write without response

dlech committed:
requirements: bump winrt version to 1.0.20330.1

We need winrt 1.0.20330.1 since it includes fixes for marshalling
Windows Guid data type to Python uuid.UUID type.

Fixes #589

dlech committed:
Merge pull request #591 from hbldh/winrt-version

requirements: bump winrt version to 1.0.20330.1

dlech committed:
v0.12.1

dlech committed:
Merge pull request #592 from hbldh/release/0.12.1

Release/0.12.1

dlech committed:
Version bump for development

dlech committed:
Merge pull request #593 from hbldh/bump

merge master back into develop

jackjansen committed:
16bit UUID support (#583)

* Allow 16bit UUIDs for get_service() and get_characteristic(), by extending the argument to a 128bit UUID.

jackjansen committed:
Register additional UUID->description mappings (#566)

* Added register_uuids() to augment the uuid-to-description mapping

Bernstern committed:
Use timeout passed into find_device_by_address in BleakScanner implementation (#600)

* Added timeout param to find_device_by_filter call

kjwill committed:
BleakClient use adapter kwarg (#608)

* BleakClient now pays attention the adapter kwarg

dlech committed:
winrt: use bleak-winrt

This switches from the Microsoft winrt package to a custom bleak-winrt
package that contains fixes that are waiting to be merged upstream.
The main fix is a major memory leak that leaks a COM object on every
await. It also allows for more efficient buffer copying.

dlech committed:
drop dotnet backend

The winrt backend seems to be working well and pythonnet still hasn't
released binary wheels for Python 3.9, so we can save ourselves some
maintenance work by only having one Windows backend.

dlech committed:
fix uart service example on Windows

Fixes the following error:

Exception in callback _ProactorReadPipeTransport._loop_reading()
handle: <Handle _ProactorReadPipeTransport._loop_reading()>
Traceback (most recent call last):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.8_3.8.2800.0_x64__qbz5n2kfra8p0\lib\asyncio\proactor_events.py", line
299, in _loop_reading
    self._read_fut = self._loop._proactor.recv(self._sock, 32768)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.8_3.8.2800.0_x64__qbz5n2kfra8p0\lib\asyncio\windows_events.py", line 445, in recv
    self._register_with_iocp(conn)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.8_3.8.2800.0_x64__qbz5n2kfra8p0\lib\asyncio\windows_events.py", line 718, in _register_with_iocp
    _overlapped.CreateIoCompletionPort(obj.fileno(), self._iocp, 0, 0)
OSError: [WinError 6] The handle is invalid

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.8_3.8.2800.0_x64__qbz5n2kfra8p0\lib\asyncio\events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.8_3.8.2800.0_x64__qbz5n2kfra8p0\lib\asyncio\proactor_events.py", line
309, in _loop_reading
    self._fatal_error(exc, 'Fatal read error on pipe transport')
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.8_3.8.2800.0_x64__qbz5n2kfra8p0\lib\asyncio\proactor_events.py", line
131, in _fatal_error
    self._force_close(exc)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.8_3.8.2800.0_x64__qbz5n2kfra8p0\lib\asyncio\proactor_events.py", line
134, in _force_close
    if self._empty_waiter is not None and not self._empty_waiter.done():
AttributeError: '_ProactorReadPipeTransport' object has no attribute '_empty_waiter'

dlech committed:
use find_device_by_filter() in uart example

This makes use of the new BleakScanner.find_device_by_filter() method to simplify the code.

dlech committed:
github: drop Python 3.6 from CI builds

The Windows backend no longer works on Python 3.6. Also the macOS build
on 3.6 had to build PyObjC from source and took a long time.

dlech committed:
Delete azure-pipelines.yml

We use GitHub actions now so this file is no longer needed.

dlech committed:
bump coverage version

We don't use this, but GitHub has advised that there is a security issue in older versions.

dlech committed:
Use correct method when writing descriptor in WinRT backend.

We always use the "with result" variant of Windows API functions to get
more information about failures. This was probably just overlooked when
porting the Windows backend to WinRT.

Fixes #615.

dlech committed:
Merge pull request #612 from hbldh/uart-example

Uart example improvements

dlech committed:
Merge pull request #616 from hbldh/fix-winrt-desc-write

Use correct method when writing descriptor in WinRT backend.

dlech committed:
Merge pull request #610 from hbldh/delete-azure-pipelines

Delete azure-pipelines.yml

dlech committed:
Merge branch 'develop' into bleak-winrt

hbldh committed:
Added `force_notify` keyword argument for WinRT backend client's `start_notify` method.

Fixes #526

elupus committed:
Only warn if handles are added again

hbldh committed:
Added `force_indicate` keyword argument for WinRT backend client's `start_notify` method.

Fixes #526

hbldh committed:
Added CHANGELOG.rst change for default notify change.

elupus committed:
Lock bluez cleanup process (#621)

Avoid race in match removal and bus disconnection by keeping lock during cleanup. Fixes #603.

Also moves reset of self.services and self._services_resolved which should fix the reconnect case of #378

dlech committed:
Merge remote-tracking branch 'origin/develop' into bleak-winrt

dlech committed:
winrt: update bleak-winrt requirement to 1.0.1

There should not be any functional changes. Main new feature Python 3.10
wheel (plus type hints).

dlech committed:
Add Python 3.10 support

dlech committed:
Merge pull request #609 from hbldh/bleak-winrt

bleak-winrt

dlech committed:
examples: use asyncio.run()

This updates all of the examples to use `asyncio.run()`. Python 3.10
issues a deprecation warning when `asyncio.get_event_loop()` is called
without creating a new event loop first.

Some other cleanups made while we are touching this:
- Use `if __name__ == "__main__":` in all examples and move command line
  argument handling there if applicable.
- Rename `run()` function to `main()` to avoid confusion with `asyncio.run()`.
- Drop use of `logging` when there was a mix of `print` and `logging`
  in the example already.

dlech committed:
drop tox.ini

This is not used and has not been maintained.

dlech committed:
drop Python 3.6 support

Python 3.6 is just a few months away from EOL and WinRT and PyObjC
already don't support it.

dlech committed:
winrt: improve error message when connecting fails

BluetoothLEDevice.FromBluetoothAddressAsync() can return null, so we
need to handle this case. The link in the comment provides some insight
in to what might cause this failure.

Fixes #647.

dlech committed:
Merge branch 'develop' into feature/issue_526

dlech committed:
winrt: default to not use cache for services

All other platforms already default to not use cache. Also, using cache
is problematic for DIY devices that don't announce that services have
changed.

Fixes #646.

dlech committed:
Merge pull request #650 from dlech/winrt-no-cache-default

winrt: default to not use cache for services

dlech committed:
Merge branch 'develop' into winrt-null-device

dlech committed:
Merge pull request #649 from dlech/winrt-null-device

winrt: improve error message when connecting fails

dlech committed:
Merge branch 'develop' into asyncio-run

dlech committed:
Merge pull request #648 from dlech/asyncio-run

examples: use asyncio.run()

dlech committed:
winrt: simplify force_indicate logic

This simplifies the logic of selecting indications or notifications
and raises an error when neither are supported instead of disabling
notifications/indications.

dlech committed:
Merge pull request #620 from hbldh/feature/issue_526

Added keyword argument option to force using Notifications instead of Indications in WinRT backend

dlech committed:
CHANGELOG: cleanup unreleased section

dlech committed:
winrt: fix protocol error descriptions

The wrong set of error messages was being used for protocol errors in
the WinRT backend. This adds a new lookup table with the correct short
error descriptions.

Fixes #532.

dlech committed:
winrt: fix notification property logic

ba424316e8ca2fade6b045e863200b37ac7ec696 was merged without proper testing and had wrong logic

dlech committed:
winrt: refactor error handling

There was quite a bit of duplicated code checking the error status
returned from GATT methods in the WinRT backend. This is refactored
into a single function the check the status and raise an exception.

Error messages are changed to use the handle instead of uuid to identify
characteristics and descriptors.

dlech committed:
Merge pull request #651 from hbldh/protocol-errors

Protocol errors

dlech committed:
winrt: add service blocklist

Trying to use the HID service on Windows causes an ACCESS_DENIED error.

This adds a list of services that should be skipped to avoid this error
when enumerating services/characteristics.

HID is currently the only known service, but more can be added as needed.

Fixes #599.

dlech committed:
Revert "winrt: default to not use cache for services"

This reverts commit 1f8f170508188537a48177a58f43c176f7395d35.

dlech committed:
winrt: remove get_services(use_cached) option

This removes the `use_cached` kwarg from `BleakClient.get_services()`
and the pass-through from `connect()` and `__init__()`. This was used
to work around issues where the cached values may be incorrect.

Since a large segment of our user base is DIY enthusiasts that trigger
this problem by changing the configuration of their DIY device, it makes
sense to just never use the cache and avoid the problem in the first
place instead of making users have to encounter the error, research it,
find the troubleshooting guide and apply the workaround.

If there is ever a legitimate use case for using cached values, we can
reconsider the feature.

dlech committed:
remove mention of "MAC" (address)

Technically, Bluetooth address are not MAC addresses. They both use a
common EUI-48 format, so people will call it a MAC address anyway, but
we will attempt to try to be slightly more correct.

dlech committed:
troubleshooting: add note on BlueZ GATT cache file

There is a bug (that has been recently fixed) in BlueZ where the GATT
cache was not removed when the device was removed, causing problems
for certain devices that change the attribute layout but don't support
the attribute database or services changed characteristics.

dlech committed:
docs: don't include _static

This is a default Sphinx config, but we currently don't have a _static
folder, so this causes an error with newer versions of Sphinx.

dlech committed:
docs: new Windows packet capture method

Microsoft has released a new tool that significantly simplifies
Bluetooth packet capture.

Fixes #653.

dlech committed:
bluez: syncronize disconnect and cleanup

When the following sequence of events occured, `assert self._bus is None`
would be hit.

- A "Connected": False property change signal is sent over D-Bus.
- Bleak handles the event and calls the `_cleanup_all()` method.
- The `_cleanup_all()` sends a D-Bus message to unsubscribe from signals and awaits it.
- While the previous method is still awaiting, the disconnect method is called.
- This method hits an assertation because the connected property is false
  but there is still a D-Bus bus connection since `_cleanup_all()` is still using it.

This fixes the issue by making `_cleanup_all()` a syncronous method.
Since we are closing the bus connection, we should not need to remove
signal watchers first.

Fixes #641.

dlech committed:
Merge pull request #661 from hbldh/windows-wireshark

Windows wireshark

dlech committed:
Merge pull request #663 from hbldh/bluez-disconnect-bug-v2

bluez: syncronize disconnect and cleanup v2

dlech committed:
docs: fix build and add CI build

This fixes some issues with building the docs and adds building docs
to CI to help prevent future regressions.

dlech committed:
Merge pull request #657 from hbldh/no-mac

add tip for BlueZ cache troubleshooting

dlech committed:
Merge pull request #656 from hbldh/winrt-cached-services

Winrt cached services

dlech committed:
Merge branch 'develop' into windows-blocked-services

dlech committed:
Merge pull request #655 from hbldh/windows-blocked-services

winrt: add service blocklist

dlech committed:
BlueZ: add lazy-get/check version method

This adds a new method to the BlueZ bakckend for checking the BlueZ
version. This lazy-gets the version from bluetoothctl so that the
subprocess only runs once.

Fixes #602.

dlech committed:
BlueZ: fix await on non-async function

This was missed in f16f40241e230c280ce9202aca760f7f33333b93

dlech committed:
Merge remote-tracking branch 'origin/develop' into pr/xloem/422

dlech committed:
p2a: fix build for recent upsteam changes

The arch argument of get_site_packages_dir() is no longer optional. Comes from changes in https://github.com/kivy/python-for-android/pull/2467

dlech committed:
examples/kivy: add .gitignore

Building from this directory as suggested in the README creates local folders that should be ignored.

dlech committed:
p4a: fix scanner discovered_devices property

This was converted from a method to a property in
https://github.com/hbldh/bleak/pull/495

dlech committed:
p4a: change is_connected from async to property

This change was made in other backends in 0f56dabc.

dlech committed:
p4a: implement client mtu_size property

This was implemented for other backends in 62f45b036f68a2.

Since Android doesn't automatically negotiate MTU, we have to do that when connecting to the device.

dlech committed:
p4a: disconnect on connection failure

If there is a failure after connecting but before the connect method returns, we need to disconnect, otherwise we end up in a bad state.

dlech committed:
p4a: raise exception on disconnect failure

returning a bool from the disconnect() method is for backwards compatilbity only and should always be True. If disconnecting failed, an exception should be raised.

dlech committed:
p4a: update example to use asyncio.run()

asyncio.run() does the same thing, but probably better.

dlech committed:
p4a: update example to use context manager

Using a context manager simplifes the code.

dlech committed:
winrt: remove get_characteristic() override

This was consolidated for all backends in ad12077a8a.

dlech committed:
Replaced usage of deprecated @abc.abstractproperty

dlech committed:
p4a: use device.name in example

Use device.name instead of scraping the name from output of __str__().

dlech committed:
p4a: fix missing overrides

several classes were missing overrides for abstract properties which causes a crash

dlech committed:
p4a: update example README

These are things I had to do to get it to build and run

dlech committed:
github: add job for building android

dlech committed:
p4a: fix return value of is_connected

This was checking disconnected instead of connected

xloem committed:
[p4a] remove shebang from android and clarify that kivy is cross-platform

xloem committed:
[p4a] use error strings from bleak.exc

dlech committed:
CoreBluetooth: add delegate methods for name/service changes

Bleak currently doesn't use these, but it could be useful to log these to see if devices in the wild actually trigger these callbacks to know if features that make use of these callbacks should be added in the future.

xloem committed:
[p4a] remove cached java constants

xloem committed:
[p4a] removed descriptor handle warning

xloem committed:
Update examples/kivy/main.py

per dlech suggestion

Co-authored-by: David Lechner <david@lechnology.com>

xloem committed:
[p4a] remove unimplemented "adapter" kwparam

xloem committed:
[p4a] iterate java lists as iterables

xloem committed:
[p4a] migrate from .format to f-strings

xloem committed:
[p4a] black formatted

xloem committed:
[p4a] do not convert connection states to strings

xloem committed:
[p4a] remove unused imports from descriptor.py

xloem committed:
Add common profile and service protocol error codes.

xloem committed:
[p4a] Use new error codes and normalise capitalisation.

xloem committed:
[p4a] use caching in github workflow

xloem committed:
[p4a] ignore android imports during autodoc

xloem committed:
[p4a] clarify how to rebuild bleak in the example

xloem committed:
[p4a] clean github action prior to build conditional on cache hit per PR recommendation

Co-authored-by: David Lechner <david@lechnology.com>

xloem committed:
Revert "[p4a] ignore android imports during autodoc"

This reverts commit 6c1e5e9dc8cde72e17c1bcd0a184a98d19bfdae2.

xloem committed:
[p4a] add android and jnius autodoc_mock_imports

xloem committed:
[p4a] scanner: initialise __adapter in __init__ for consistency

dlech committed:
CHANGELOG: spelling

dlech committed:
CHANGELOG: change heading style

Using `---` instead of `~~~` makes it trivial to copy the relevant
section of the changelog to github releases and have nice formatting.

dlech committed:
Merge pull request #665 from hbldh/CBPeripheralDelegate-extra-methods

CoreBluetooth: add delegate methods for name/service changes

dlech committed:
Merge pull request #664 from hbldh/bluez-version

BlueZ: add lazy-get/check version method

dlech committed:
CoreBluetooth: remove use_cached option

This removes the use_cached parameter when enumerating services/
characteristics/descriptors. Services would incorrectly return None
when reconnecting due to using `self._services_discovered_future.done()`
instead of `self.peripheral.discoverServices_(None) is not None`.

We should always re-enumerate the services after reconnecting, so
having use_cached doesn't make sense.

dlech committed:
change from get_event_loop() to get_running_loop()

Since we have dropped support for Python 3.6, we can now use the
preferred `asyncio.get_running_loop()` instead of
`asyncio.get_event_loop()`.

dlech committed:
fix doc build

readthedocs only supports up to Python 3.8

dlech committed:
Merge pull request #672 from hbldh/get_running_loop

change from get_event_loop() to get_running_loop()

dlech committed:
Merge pull request #667 from hbldh/mac-no-cache

CoreBluetooth: remove use_cached option

dlech committed:
Merge branch 'develop' into pr/xloem/422

dlech committed:
p4a: remove CONTROLLER_ERROR_CODES from error code table

The values in CONTROLLER_ERROR_CODES overlap with PROTOCOL_ERROR_CODES
so we will need a separate lookup table if these error codes are
actually needed.

dlech committed:
p4a: remove unused import

dlech committed:
p4a: spelling

dlech committed:
p4a: convert SCAN_FAILED_NAMES to IntEnum

dlech committed:
p4a: use bytes in advertising data

To be like other backends, use bytes instead of bytearray for service
and manufacturer data. Also simplify code a bit while we are touching
this.

dlech committed:
Merge pull request #422 from xloem/p4a-beta

python-for-android backend

dlech committed:
v0.13.0

Added
-----

* Allow 16-bit UUID string arguments to ``get_service()`` and ``get_characteristic()``.
* Added ``register_uuids()`` to augment the uuid-to-description mapping.
* Added support for Python 3.10.
* Added ``force_indicate`` keyword argument for WinRT backend client's ``start_notify`` method. Fixes #526.
* Added python-for-android backend.

Changed
-------

* Changed from ``winrt`` dependency to ``bleak-winrt``.
* Improved error when connecting to device fails in WinRT backend. Fixes #647.
* Changed examples to use ``asyncio.run()``.
* Changed the default notify method for the WinRT backend from ``Indicate`` to ``Notify``.
* Refactored GATT error handling in WinRT backend.
* Changed Windows Bluetooth packet capture instructions. Fixes #653.
* Replaced usage of deprecated ``@abc.abstractproperty``.
* Use ``asyncio.get_running_loop()`` instead of ``asyncio.get_event_loop()``.

Removed
-------

* Removed ``dotnet`` backend.
* Dropped support for Python 3.6.
* Removed ``use_cached`` kwarg from ``BleakClient`` ``connect()`` and ``get_services()`` methods. Fixes #646.

Fixed
-----

* Fixed unused timeout in the implementation of BleakScanner's ``find_device_by_address()`` function.
* Fixed BleakClient ignoring the ``adapter`` kwarg. Fixes #607.
* Fixed writing descriptors in WinRT backend. Fixes #615.
* Fixed race on disconnect and cleanup of BlueZ matches when device disconnects early. Fixes #603.
* Fixed memory leaks on Windows.
* Fixed protocol error code descriptions on WinRT backend. Fixes #532.
* Fixed race condition hitting assentation in BlueZ ``disconnect()`` method. Fixes #641.
* Fixed enumerating services on a device with HID service on WinRT backend. Fixes #599.
* Fixed subprocess running to check BlueZ version each time a client is created. Fixes #602.
* Fixed exception when discovering services after reconnecting in CoreBluetooth backend.

dlech committed:
Merge pull request #622 from elupus/service_warn

Only warn if handles are added again

dlech committed:
CHANGELOG: add entry from #622

dlech committed:
Merge branch 'develop' into release/0.13.0

dlech committed:
CHANGELOG: update release date

dlech committed:
Merge pull request #673 from hbldh/release/0.13.0

Release/0.13.0

dlech committed:
setup: add project URLs

The changelog is removed from the long description because it was
causing an error when uploading to pypi. Additional URLs are helpful
too.

dlech committed:
bump version for develop

dlech committed:
setup: add Android classifier

we have Android support now.

dlech committed:
setup: change topic

This is a non-standard topic, but communications is for things like
email and fax, so it doesn't seem like the right fit. There is already
a similar "Topic :: System :: Hardware :: Universal Serial Bus (USB)".

dlech committed:
CoreBluetooth: delete futures after awaiting

In the CoreBluetooth backend, futures are used to pass results/errors
from the delegate callbacks to methods that are waiting for them. The
futures are stored as part of the object state so that the delegates
can access them.

Prior to this change, completed futures were not removed from the
object status. This worked most of the time because a second call
of the same method would replace the completed future from the prior
call. However, in the case of notifications and reads, the same
delegate callback is shared with two methods. So when both of these
were used, notifications would attempt to complete the already completed
read future which raises an InvalidStateError.

Fixes #675.

dlech committed:
CoreBluetooth: pass return values via futures

This updates CoreBluetooth to read the changed value in the delegate
callback and pass it back to the BleakClient method via the future
object.

By reading the value in the delegate callback rather than the method
scheduled by call_soon_threadsafe(), we ensure that we get the original
value from the delegate callback and not a possibly changed value later.

dlech committed:
CoreBluetooth: drop return True in PeripheralDelegate

This removes `return True` from several PeripheralDelegate methods.
These methods unconditionally return `True`, so we can simplify the
code by removing it.

dlech committed:
CoreBluetooth: only notify if not read pending

The same delegate callback is used for notifications and reading
characteristics. This rearranges the logic so that if there is a
read request pending, it will complete the read, otherwise it will
trigger a notification callback. Previously, reads would also trigger
notification callbacks.

dlech committed:
Merge pull request #678 from hbldh/cb-futures

CoreBluetooth callback fixes and cleanups

dlech committed:
winrt: use GATT session status

This changes the WinRT backend to use the GATT session status instead
of the actual device connection status. Windows doesn't actually allow
explicitly connecting and disconnecting devices. This is handled
automatically in the background. So instead, we use the GATT session
to act as the connection status. This could potentially fix some issues
where devices disconnect briefly after initial connection and before
enumerating services.

dlech committed:
Merge pull request #680 from hbldh/winrt-gatt-session

winrt: use GATT session status

dlech committed:
add black to Pipfile

dlech committed:
corebluetooth: work around macOS 12 scanner bug

macOS has a bug (feature?) where advertising data is no longer received unless one
or more service UUIDs are given in scanForPeripheralsWithServices:options:.

This implements a new kwarg on `BleakScanner` to allow users to provide
such a list of UUIDs. This commit only implements it in the CoreBluetooth
backend with other backends to follow.

Issue #635.

dlech committed:
bluezdbus: implement service_uuids kwarg

This implements the service_uuids kwarg in BleakScanner in the BlueZ backend.

dlech committed:
p4android: set_scanning_filter is not implemented

This changes set_scanning_filter in the Android backend to raise
NotImplementedError since it is not implemented. Also remove async
specifier since this is not supposed to be an async function.

dlech committed:
p4android: implement BleakScanner service_uuids kwarg

This implements the BleakScanner service_uuids kwarg in the Android backend.

dlech committed:
winrt: combine adv data and scan rsp data

This changes how advertising data is cached and combined. Previously,
we just cached scan response data to get the local name later. Separate
callbacks were called for advertising data and scan response data which
makes more work for users who are interested in filtering on data that
is split between the two packets.

Now, the advertising data and scan response data are both cached together
as a single object and this object is used to create the Bleak advertising
data structure so that the advertising data seen by Bleak users now
contains the information from both advertising data type combined.

dlech committed:
winrt: implement BleakScanner service_uuids kwarg

This provides an implementation for the BleakScanner service_uuids
kwarg in the WinRT backend. As explained in the code comment, it is
not possible to use the OS-provided filter mechanism for technical
reasons.

Fixes #230

dlech committed:
corebluetooth: fix crash on disconnect

Fixes the following exception:

    Traceback (most recent call last):
    File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/events.py", line 80, in _run
        self._context.run(self._callback, *self._args)
    File "/Users/david/work/bleak/bleak/backends/corebluetooth/CentralManagerDelegate.py", line 358, in did_disconnect_peripheral
        callback()
    File "/Users/david/work/bleak/bleak/backends/corebluetooth/client.py", line 105, in disconnect_callback
        for future in self._delegate.futures():
    File "/Users/david/work/bleak/bleak/backends/corebluetooth/PeripheralDelegate.py", line 77, in futures
        if self._services_discovered_future
    AttributeError: 'PeripheralDelegate' object has no attribute '_services_discovered_future'

Problem was introduced by commit ad59f6e8e35cbab8276d654faafa7856c1fd8b9b.

dlech committed:
Merge pull request #696 from hbldh/fix-mac-disconnect

corebluetooth: fix crash on disconnect

dlech committed:
Merge pull request #692 from hbldh/macos12-scanning

corebluetooth: work around macOS 12 scanner bug

mkoderer committed:
Use predefined service uuids

On macos the service_uuids need to be filled. Main issue for
beginners is to retrieve some valid uuids. Instead of using an
external tool like nRF we could use predefined ones in bleak/uuids.

Implementation works with 128bit and 16bit UUIDs.

Related-to #635

Signed-off-by: Marc Koderer <marc@koderer.com>

dlech committed:
Merge pull request #706 from mkoderer/all_uuids

Use predefined service uuids

dlech committed:
uuids: add LEGO UUIDs

LEGO has Powered Up Smart Hubs that communicate with BLE. This adds the
UUIDs for the official LEGO firmware and 3rd-party Pybricks firmware.

dlech committed:
requirements: bump bleak-winrt to v1.1.0

This removes the explicit apartment initialization that breaks apps
that have already initialized a single threaded apartment.

Fixes: https://github.com/hbldh/bleak/issues/698

dlech committed:
github: convert android build to manual trigger

The Android build is failing and I don't have time to fix it. This
converts it to a manual trigger so that it doesn't cause pull requests
to go red but keeps the job so we can fix it later.

dlech committed:
drop pytest-runner

Apparently there are some security concerns about pytest-runner. We
weren't using it anyway, so we can remove it.

dlech committed:
Merge pull request #711 from hbldh/drop-pytest-runner

drop pytest-runner

dlech committed:
Merge pull request #710 from hbldh/android-build

github: convert android build to manual trigger

dlech committed:
Merge pull request #709 from hbldh/bleak-winrt

requirements: bump bleak-winrt to v1.1.0

dlech committed:
Merge pull request #708 from hbldh/lego

uuids: add LEGO UUIDs

dlee1j1 committed:
small bug fix to set the metadata on the device on discovery callback (#715)

dlech committed:
Limit "all" option to Mac. (#717)

Adding 100s of UUIDs to the scan filter on BlueZ causes BlueZ to fail
with:

    Traceback (most recent call last):
    File "examples/detection_callback.py", line 53, in <module>
        asyncio.run(main(service_uuids))
    File "/usr/lib/python3.8/asyncio/runners.py", line 44, in run
        return loop.run_until_complete(main)
    File "/usr/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
        return future.result()
    File "examples/detection_callback.py", line 42, in main
        await scanner.start()
    File "/home/david/work/pybricks/bleak/bleak/backends/bluezdbus/scanner.py", line 169, in start
        assert_reply(reply)
    File "/home/david/work/pybricks/bleak/bleak/backends/bluezdbus/utils.py", line 23, in assert_reply
        raise BleakDBusError(reply.error_name, reply.body)
    bleak.exc.BleakDBusError: [org.bluez.Error.InProgress] Operation already in progress

`bluetoothd` has to be restarted to clear the error.

So don't allow this option on OS other than Mac. Also drop the "all"
option and enable it by default based on the macOS version.

Fixes: https://github.com/hbldh/bleak/issues/716

tcamise-gpsw committed:
Change logging.error to logger.error (#726)

dlech committed:
Version 0.14.0

Added
-----

* Added ``service_uuids`` kwarg to  ``BleakScanner``. This can be used to work around issue of scanning not working on macOS 12. Fixes #230. Works around #635.
* Added UUIDs for LEGO Powered Up Smart Hubs.

Changed
-------

* Changed WinRT backend to use GATT session status instead of actual device connection status.
* Changed handling of scan response data on WinRT backend. Advertising data and scan response data is now combined in callbacks like other platforms.
* Updated ``bleak-winrt`` dependency to v1.1.0. Fixes #698.

Fixed
-----

* Fixed ``InvalidStateError`` in CoreBluetooth backend when read and notification of the same characteristic are used. Fixes #675.
* Fixed reading a characteristic on CoreBluetooth backend also triggers notification callback.
* Fixed in Linux, scanner callback not setting metadata parameters. Merged #715.

dlech committed:
setup.py: drop Bluetooth classifier

PyPI rejects this classifier.

dlech committed:
github: change release trigger

Using the published trigger is more robust than the created trigger.
It allows triggering the workflow from a draft (helpful when a release
upload fails and you have to do it over again).

dlech committed:
Merge pull request #727 from hbldh/release/0.14.0

Release/0.14.0

dlech committed:
Bump version for develop.

dlech committed:
Bump version for develop.

dlech committed:
Bump version for develop.

dlech committed:
flake8: ignore more generated files

dlech committed:
corebluetooth: don't log scanner error in app

If bleak is running in a .app (e.g. created with py2app), then scanning
in macOS 12 works as it did in previous versions, so we don't need to
log the error. We can determine this by checking the bundle id since
Python itself is a .app and has a bundle id.

Related: https://github.com/hbldh/bleak/issues/720

dlech committed:
docs: warn against bleak.py

We frequently get bug reports about circular import errors. This is
due to naming your script bleak.py. So put warnings everywhere so
hopefully people won't miss it.

Fixes: #729
Fixes: #704
Fixes: #515

dlech committed:
winrt: fix AttributeError in BleakClient.__init__()

This fixes a regression caused by 3f26be6. When a `BleDevice` is passed
instead of a string, it tries to loop up a non-existent attribute.

Fixes: #731

dlech committed:
Merge pull request #730 from hbldh/warn-bleak-py

docs: warn against bleak.py

dlech committed:
Merge pull request #732 from hbldh/winrt-fix-regression

winrt: fix AttributeError in BleakClient.__init__()

dlech committed:
v0.14.1

dlech committed:
Merge pull request #735 from hbldh/release/0.14.1

Release/0.14.1

dlech committed:
winrt: update bleak-winrt to v1.1.1

v1.1.0 was not statically linked against the MSVC runtime and would
require users to manually install it to avoid an ImportError.

Fixes: https://github.com/hbldh/bleak/issues/741

dlech committed:
Merge pull request #742 from hbldh/bleak-winrt-1.1.1

winrt: update bleak-winrt to v1.1.1

dlech committed:
winrt: fix name is 'Unknown'

Commit 3f26be60c "winrt: combine adv data and scan rsp data" incorrectly
assumed that `local_name` would return None instead of '' when no name
was present in the advertising data. This changes the test to not
strictly check for None.

Fixes: https://github.com/hbldh/bleak/issues/736

dlech committed:
Merge pull request #748 from hbldh/winrt-fix-name-unknown

winrt: fix name is 'Unknown'

dlech committed:
v0.14.2

dlech committed:
Merge pull request #749 from hbldh/release/0.14.2

v0.14.2

dlech committed:
README: add download stats

Commit:
bluezdbus: Return the battery percentage in an array

bytearray(val) will create a zero-filled instance of length val.
bytearray([val]) will create a bytearray containing val.

dlech committed:
Merge pull request #751 from quadstew/fix-bluez-battery

bluezdbus: Return the battery percentage in an array

dlech committed:
Merge pull request #728 from hbldh/macos-background-bug

Skip logging CoreBlueooth scanner error if running as .app

dlech committed:
format with black 22.1.0

Black is stable now, so we shouldn't have to do this again for a whole year.

Commit:
added local device info object for pair/unpair

dlech committed:
Merge pull request #757 from harshdeepsehgal/develop

Local device info object for pair/unpair

pmessan committed:
fix(scanner): Add suspend and resume to scanner

pmessan committed:
Merge remote-tracking branch 'origin/proglove' into proglove-update

pmessan committed:
Merge remote-tracking branch 'origin/proglove'
into proglove-update

arthur-proglove commented:
There are 776 commits and 129 files changed, what am I supposed to review here?

I would suggest to take the upstream from Bleak, apply the patches you think might be relevant and create a PR from there.

pmessan commented:
> There are 776 commits and 129 files changed, what am I supposed to review here?
> 
> I would suggest to take the upstream from Bleak, apply the patches you think might be relevant and create a PR from there.

That's actually what I did. I rebased this branch on top of the upstream develop branch, which had had a lot of commits since the `proglove` branch was forked. What I wanted you to check was the last 3 commits I added, and to verify if the patches I added are okay. My thinking was that since we are several versions behind, there would be a lot of conflicts with the base branch, so we opt in favor of the upstream branch if its shown to not break our code in `progateway` and `problew`.
